### PR TITLE
Dinit support

### PR DESCRIPTION
**Pull request by MagicRB**

### CODE REVIEW

**Comment by pizzapim** [comment]
Have pushed some fixes I found so far here: https://github.com/pizzapim/NixNG/tree/dinit-support

**Comment by MagicRB** [comment]
Well, i patched it in, it compiles, ill test later

**Comment by MagicRB** [comment]
@pizzapim i merged your changes and added a multiple `env-file` patch to `dinit`

**Comment by MagicRB** [comment]
I rebased, though maybe i broke something

**Comment by pizzapim** [comment]
Two things I'm noticing now:
1. Apparently the `chpst` from `busybox` does not implement the `-b` flag, so those instances would have to be fixed if we swap to dinit
2. `runit` still seems to be used somewhere even if you enable `dinit`

**Comment by MagicRB** [comment]
1. Well most of the modules use the `runit` `chpst` i think, so that should be fine
2. See 1

**Comment by pizzapim** [comment]
I mean, `runit` is still being added to `environment.systemPackages` in `modules/environment.nix`. I don't think this is necessary at all?

**Comment by MagicRB** [comment]
Oh yeah that's an oversight. The actual runit module should do that conditionally only, it really shouldn't be present always.

**Comment by pizzapim** [comment]
Also, seems like `dumb-init` has a hidden dependency on `runit` because it expects `system.activation."runit"` to be ran.

File: overlay/dinit.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  src = fetchFromGitHub {
+    owner = "davmac314";
+    repo = "dinit";
+    rev = "v" + self.version;
+    hash = "sha256-ApB0pEFSyawNASF/rqRmhT4FLofZzYmNdNmG2FGpnnk=";
+  };
+
+  patches = [
+    ./dinit-dont-shutdown-system.patch
```
[the remark this answers is not in the document]
Comment (MagicRB) [reply overlay/dinit.nix:19]: Thing is we want it to do process reaping and properly waiting on services to come down, but not call the kernel to shutdown. So a yes and no situation

[the remark this answers is not in the document]
Comment (aanderse) [reply overlay/dinit.nix:19]: yeah, i _think_ that is what `-s` does...

```
 --system, -s                 run as the system service manager
 --system-mgr, -m             run as system manager (perform shutdown etc)
```

[the remark this answers is not in the document]
Comment (MagicRB) [reply overlay/dinit.nix:19]: We've tested that itll try ro do shutdown without that patch and we do use `-s` so im a bit confused in regards to the docs

[the remark this answers is not in the document]
Comment (aanderse) [reply overlay/dinit.nix:19]: in my experiments i ran an instance of `dinit` during stage1, shut it down, then launched a `dinit` again for stage2 and everything worked as i expected it to

from my perspective it worked as i am describing

[the remark this answers is not in the document]
Comment (MagicRB) [reply overlay/dinit.nix:19]: Ok, ill retest during the weekend and finish this PR up. Thanks for input

[the remark this answers is not in the document]
Comment (MagicRB) [reply overlay/dinit.nix:19]: @aanderse
```
[STOPPD] nginx
Sending TERM/KILL to all processes...
2024/12/13 15:20:25 [notice] 100#100: signal 15 (SIGTERM) received from 1, exiting
2024/12/13 15:20:25 [notice] 100#100: signal 17 (SIGCHLD) received from 101
2024/12/13 15:20:25 [notice] 100#100: worker process 101 exited with code 0
2024/12/13 15:20:25 [notice] 100#100: signal 29 (SIGIO) received
2024/12/13 15:20:25 [notice] 100#100: signal 17 (SIGCHLD) received from 102
2024/12/13 15:20:25 [notice] 100#100: worker process 102 exited with code 0
2024/12/13 15:20:25 [notice] 100#100: exit
Turning off swap...
/sbin/swapoff: No such file or directory
Unmounting disks...
/bin/umount: No such file or directory
Issuing shutdown via kernel...
```
that's what I get without the patch

[the remark this answers is not in the document]
Comment (MagicRB) [reply overlay/dinit.nix:19]: Problem is we need this bit to execute https://github.com/davmac314/dinit/blob/master/src/shutdown.cc#L434 but not the shutdown, i.e. we need process reaping but not shutdown. Without the process reaping dinit exits immediately and then the container runtime will `SIGKILL` the rest of the processes.

**Comment by MagicRB** [comment]
I'm gonna merge this, it seems to be good enough and it being in a branch is making my life more difficult in my dotfiles repo.

**Pull request merged**